File: libs/graphicsgeo/src/main/kotlin/mono/graphics.geo/Point.kt
package mono.graphics.geo

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

@Serializable(with = Point.PointSerializer::class)
data class Point(val left: Int, val top: Int) {
    val row: Int get() = top
    val column: Int get() = left

    operator fun minus(base: Point): Point = Point(left - base.left, top - base.top)

    operator fun plus(base: Point): Point = Point(left + base.left, top + base.top)

    internal object PointSerializer : KSerializer<Point> {
        override val descriptor: SerialDescriptor =
            PrimitiveSerialDescriptor("Point", PrimitiveKind.STRING)

        override fun serialize(encoder: Encoder, value: Point) {
            encoder.encodeString("${value.left}|${value.top}")
        }

        override fun deserialize(decoder: Decoder): Point {
            val marshaledValue = decoder.decodeString()
            val (left, top) = marshaledValue.split("|")
            return Point(left = left.toInt(), top = top.toInt())
        }
    }

    companion object {
        val ZERO = Point(0, 0)
    }
}

@Serializable(with = DirectedPoint.DirectedPointSerializer::class)
data class DirectedPoint(val direction: Direction, val left: Int, val top: Int) {
    val point: Point
        get() = Point(left, top)

    constructor(direction: Direction, point: Point) : this(direction, point.left, point.top)

    operator fun plus(base: Point): DirectedPoint =
        copy(left = left + base.left, top = top + base.top)

    enum class Direction {
        HORIZONTAL,
        VERTICAL;

        val normalizedDirection: Direction
            get() = when (this) {
                VERTICAL -> HORIZONTAL
                HORIZONTAL -> VERTICAL
            }
    }

    internal object DirectedPointSerializer : KSerializer<DirectedPoint> {
        private const val MARSHAL_HORIZONTAL = "H"
        private const val MARSHAL_VERTICAL = "V"

        override val descriptor: SerialDescriptor =
            PrimitiveSerialDescriptor("DirectedPoint", PrimitiveKind.STRING)

        override fun serialize(encoder: Encoder, value: DirectedPoint) {
            val direction = when (value.direction) {
                Direction.HORIZONTAL -> MARSHAL_HORIZONTAL
                Direction.VERTICAL -> MARSHAL_VERTICAL
            }
            encoder.encodeString("$direction|${value.left}|${value.top}")
        }

        override fun deserialize(decoder: Decoder): DirectedPoint {
            val marshaledValue = decoder.decodeString()
            val (marshaledDirection, left, top) = marshaledValue.split("|")
            val direction =
                if (marshaledDirection == MARSHAL_HORIZONTAL) {
                    Direction.HORIZONTAL
                } else {
                    Direction.VERTICAL
                }
            return DirectedPoint(
                direction,
                left = left.toInt(),
                top = top.toInt()
            )
        }
    }
}

File: monosketch-svelte/src/lib/mono/graphics-geo/point.ts
import type { Comparable } from '$mono/base-interface/comparable';

export class Point implements Comparable {
    static readonly ZERO = new Point(0, 0);

    static of(left: number, top: number): Point {
        return new Point(left, top);
    }

    static from(point: { left: number; top: number }): Point {
        return new Point(point.left, point.top);
    }

    constructor(
        public readonly left: number,
        public readonly top: number,
    ) {}

    get row(): number {
        return this.top;
    }

    get column(): number {
        return this.left;
    }

    equals(other: unknown): boolean {
        if (!(other instanceof Point)) {
            return false;
        }
        return this.left === other.left && this.top === other.top;
    }

    /**
     * Returns a new point that is the difference between this point and the
     * other point.
     */
    minus(other: Point): Point {
        return new Point(this.left - other.left, this.top - other.top);
    }

    /**
     * Returns a new point that is the sum of this point and the other point.
     */
    plus(other: Point): Point {
        return new Point(this.left + other.left, this.top + other.top);
    }

    toString(): string {
        return `${this.left}|${this.top}`;
    }

    // TODO: implement serialize and deserialize
}

export enum Direction {
    HORIZONTAL,
    VERTICAL,
}

export class DirectedPoint implements Comparable {
    static of(direction: Direction, left: number, top: number): DirectedPoint {
        return new DirectedPoint(direction, new Point(left, top));
    }

    constructor(
        public readonly direction: Direction,
        public readonly point: Point,
    ) {}

    get left(): number {
        return this.point.left;
    }

    get top(): number {
        return this.point.top;
    }

    equals(other: unknown): boolean {
        if (!(other instanceof DirectedPoint)) {
            return false;
        }
        return this.direction === other.direction && this.point.equals(other.point);
    }

    /**
     * Returns a new directed point that has the point is sum of this point and
     * the base point.
     */
    plus(base: Point): DirectedPoint {
        return new DirectedPoint(this.direction, this.point.plus(base));
    }

    // TODO: implement serialize and deserialize
}
